### PR TITLE
blockstore: make it safe for concurrent use

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

### CODE REVIEW

File: v2/blockstore/readonly.go [hunk 1]
```diff
@@ -152,6 +170,9 @@ func (b *ReadOnly) PutMany([]blocks.Block) error {
 
 // AllKeysChan returns the list of keys in the CAR.
 func (b *ReadOnly) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
```
Comment (dirkmc) [inline v2/blockstore/readonly.go:175]: Do we also need to take the read lock inside the go-routine?

**Review by dirkmc — COMMENTED** [review]
Do we need to take a lock in the Roots and Close methods?

File: v2/blockstore/readonly.go (hunk 1 above)
Comment (mvdan) [reply v2/blockstore/readonly.go:175]: Yikes, good catch. We probably want to grab the mutex here, and release it when the other goroutine ends.

Comment (mvdan) [reply v2/blockstore/readonly.go:175]: Done.

**Comment by aarshkshah1992** [comment]
LGTM. Just need to address @dirkmc 's comment here.

**Comment by mvdan** [comment]
It's already addressed, so merging.

**Comment by dirkmc** [comment]
@mvdan do we need to take a lock in the Roots and Close methods?

**Comment by mvdan** [comment]
Neither is part of the blockstore interface, so my intuition is that they don't need locking. Close should never be called concurrently to reads, as far as I can tell. And the roots are set at the beginning and never modified, so that should never be racy.

Really, the only racy bits here are the index and the io.Writer for appending blocks to the carv1.

**Pull request merged**